### PR TITLE
Replace home screen's UITableView to UICollectionView

### PR DESCRIPTION
**Pull request by jaewonsim**
- Replaces home screen's UITableView to UICollectionView
- Updates layout for homeCollectionView accordingly
- Fix minor UI mistakes in home screen's sub-UICollectionViews' cells

### CODE REVIEW

File: Recast/HomeViewController.swift [hunk 1]
```diff
@@ -30,24 +30,36 @@ enum HomeSectionType: String, CaseIterable {
...
     ///
     /// - Parameters:
-    ///   - identifier: Indicates if `value` is the `int` for `section` or `tag`
-    ///   - value: The `int` value of the `section` or `tag`
-    /// - Returns: The HomeSectionType for the corresponding tag or section value of the HomeTableView
-    static func type(for identifier: IdentifierType, value: Int) -> HomeSectionType? {
-        var section: Int? = value
+    ///   - identifier: Indicates if `value` is the `tag` or `homeCollectionViewSection`
+    ///   - value: The `tag` or `homeCollectionViewSection`
+    /// - Returns: The HomeSubsection for the corresponding tag or section
```
Comment (mindylou) [inline Recast/HomeViewController.swift:50]: wow this documentation is next level 👌

File: Recast/HomeViewController.swift [hunk 2]
```diff
@@ -56,6 +68,14 @@ enum HomeSectionType: String, CaseIterable {
 
 class HomeViewController: UIViewController {
 
+    /// Returns if the given indexPath section is that of the last section in the homeCollectionView
+    ///
+    /// - Parameter section: indexPath section to determine for
+    /// - Returns: if the given indexPath section is that of the last section in the homeCollectionView
+    func isLastSection(section: Int) -> Bool {
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
    func isLastSection(_ section: Int) -> Bool {
```

File: Recast/HomeViewController.swift [hunk 3]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-        return 60
+        if let subsection = HomeSubsection.homeSubsection(for: .tag, value: collectionView.tag) {
+            switch subsection {
+            case .yourFavorites, .browseTopics:
+                return gridItemSize
+            case .continueListening:
+                return continueListeningItemSize
+            }
+        } else {
+            return CGSize.zero
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
            return .zero
```

File: Recast/HomeViewController.swift [hunk 4]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-}
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
 
-// MARK: - TV Data Source
-extension HomeViewController: UITableViewDataSource {
-    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return 1
-    }
+        let homeCollectionViewSectionInset = UIEdgeInsets(top: 0, left: 22, bottom: 0, right: 22)
+        let subCollectionViewSectionInset = UIEdgeInsets(top: 0, left: 22, bottom: 0, right: 22)
```
Comment (mindylou) [inline Recast/HomeViewController.swift]: consider keeping a constant in this function with the left + right insets

File: Recast/HomeViewController.swift [hunk 5]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-            cell.collectionView.reloadData()
-            return cell
-        default:
-            return UITableViewCell(frame: .zero)
+        if collectionView == homeCollectionView, isLastSection(section: section) {
+            return homeCollectionViewSectionInset
+        } else if HomeSubsection.homeSubsection(for: .tag, value: collectionView.tag) != nil {
+            return subCollectionViewSectionInset
+        } else {
+            return UIEdgeInsets.zero
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
            return .zero
```

File: Recast/HomeViewController.swift [hunk 6]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-            break
-        case .browseTopics:
-            break
+// MARK: - CV Data Source
+extension HomeViewController: UICollectionViewDataSource {
+
+    func numberOfSections(in collectionView: UICollectionView) -> Int {
+        if collectionView == homeCollectionView {
+            return HomeSubsection.allCases.count + 1 //+1 accounts for the last section for subscriptions
+        } else {
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
```

File: Recast/HomeViewController.swift [hunk 7]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-}
 
-// MARK: - CV Data Source
-extension HomeViewController: UICollectionViewDataSource {
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        let type = HomeSectionType.type(for: .tag, value: collectionView.tag)!
+        if collectionView == homeCollectionView {
+            if isLastSection(section: section) {
+                return subscriptions.count
+            } else {
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
```

File: Recast/HomeViewController.swift [hunk 8]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-            cell.collectionView.tag = viewTag
-            cell.collectionView.reloadData()
-            return cell
-        default:
-            return UITableViewCell(frame: .zero)
+        if collectionView == homeCollectionView, isLastSection(section: section) {
+            return homeCollectionViewSectionInset
+        } else if HomeSubsection.homeSubsection(for: .tag, value: collectionView.tag) != nil {
+            return subCollectionViewSectionInset
+        } else {
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
```

File: Recast/HomeViewController.swift [hunk 9]
```diff
@@ -157,92 +204,81 @@ class HomeViewController: UIViewController {
...
-    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
-        return 60
+        if let subsection = HomeSubsection.homeSubsection(for: .tag, value: collectionView.tag) {
+            switch subsection {
+            case .yourFavorites, .browseTopics:
+                return gridItemSize
+            case .continueListening:
+                return continueListeningItemSize
+            }
+        } else {
```
Comment (mindylou) [inline Recast/HomeViewController.swift]:
```suggestion
```

File: Recast/HomeViewController.swift [hunk 10]
```diff
@@ -253,43 +289,96 @@ extension HomeViewController: UICollectionViewDataSource {
...
+            let header = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: homeHeaderReuse, for: indexPath) as! HomeCollectionViewHeader
+            if isLastSection(section: indexPath.section) {
+                header.headerTitleLabel.text = "Subscriptions"
+            } else {
+                header.headerTitleLabel.text = HomeSubsection.allCases[indexPath.section].rawValue
+            }
+            return header
+
+        default:
+            assert(false)
```
Comment (mindylou) [inline Recast/HomeViewController.swift]: not sure if we want an assert here bc it'll break the app, maybe return an empty `UICollectionReusableView`?

**Review by mindylou — CHANGES_REQUESTED** [review]
great work so far!! just a couple of comments and suggested changes

File: Recast/Cells/HomeContinueListeningCollectionViewCell.swift [hunk 11]
```diff
@@ -9,25 +9,22 @@
...
+/// UICollectionViewCell subclass to be used for displaying a "Continue Listening" collection view inside a table view
+//swiftlint:disable:next type_name
+class HomeContinueListeningCollectionViewCell: UICollectionViewCell {
     var collectionView: UICollectionView!
     private let continueCvReuse = "continueListeningCvReuse"
 
     // MARK: Lifecycle
-    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
-        super.init(style: style, reuseIdentifier: reuseIdentifier)
+    override init(frame: CGRect) {
```
Comment (doctor3w) [inline Recast/Cells/HomeContinueListeningCollectionViewCell.swift:19]: Does CollectionViewCell not have a reuseIdentifier init? If it does we should use that.

File: Recast/Cells/HomePodcastGridCollectionViewCell.swift [hunk 12]
```diff
@@ -8,27 +8,24 @@
...
     // MARK: View vars
     var collectionView: UICollectionView!
 
     // MARK: CV reuse identifier
     private let seriesGridReuse = "gridCvReuse"
 
     // MARK: Lifecycle
-    override init(style: UITableViewCellStyle, reuseIdentifier: String?) {
-        super.init(style: style, reuseIdentifier: reuseIdentifier)
+    override init(frame: CGRect) {
```
Comment (doctor3w) [inline Recast/Cells/HomePodcastGridCollectionViewCell.swift:20]: Same here

File: Recast/HomeViewController.swift [hunk 13]
```diff
@@ -157,92 +204,76 @@ class HomeViewController: UIViewController {
...
-    func homeTableViewHeaderView(_ tableHeader: HomeTableViewHeaderView, didPress: Action) {
-        if tableHeader == headerView {
-            switch didPress {
-            case .seeAll:
-                print("Pressed see all")
-            }
+// MARK: - CV Delegate
+extension HomeViewController: UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
+        let type = HomeSubsection.homeSubsection(for: .tag, value: collectionView.tag)!
```
Comment (doctor3w) [inline Recast/HomeViewController.swift:210]: Does this cause swiftlint warnings?

**Review by doctor3w — COMMENTED** [review]
This looks awesome! Great work! Just a few questions!

**Review by mindylou — DISMISSED** [review]

File: Recast/HomeViewController.swift [hunk 14]
```diff
@@ -30,24 +30,36 @@ enum HomeSectionType: String, CaseIterable {
         }
     }
 
-    /// The HomeSectionType for the corresponding tag or section value of the HomeTableView.
+    /// Returns for the row of `homeCollectionView` that corresponds to the `HomeSubsection`
+    var homeCollectionViewSection: Int? {
```
Comment (jackthmp) [inline Recast/HomeViewController.swift:34]: I feel like this can be replaced with using HomeSubsection.allCases

**Review by jackthmp — DISMISSED** [review]

**Review by mindylou — APPROVED** [review]

**Review by jackthmp — APPROVED** [review]

**Pull request merged**